packages/minimongo/local_collection.js

```diff
@@ -669,17 +669,18 @@ LocalCollection._CachingChangeObserver = class _CachingChangeObserver {
       this.docs = new OrderedDict(MongoID.idStringify);
       this.applyChange = {
         addedBefore: (id, fields, before) => {
-          const doc = EJSON.clone(fields);
+          // Take a shallow copy since the top-level properties can be changed
+          const doc = { ...fields };
 
           doc._id = id;
 
           if (callbacks.addedBefore) {
-            callbacks.addedBefore.call(this, id, fields, before);
+            callbacks.addedBefore.call(this, id, EJSON.clone(fields), before);
```

Comment:
Hm, why are you here both doing the shallow clone and still EJSON.cloning fields?

Reply:
There are 2 things to make sure of:
1. We don't modify the object we get in and save to the `docs` of `CachingChangeObserver`. Therefore we take a shallow clone. Before this, this was not guaranteed by the CachingChangeObserver, so it was always doing a deep clone outside, e.g.:
https://github.com/meteor/meteor/blob/devel/packages/mongo/observe_multiplex.js#L162
But this is not necessary. At least as long as you stick to the contract a shallow clone is sufficient https://github.com/meteor/meteor/blob/d072b5b43256005dcb3d96dead5474c1f4893a8d/packages/minimongo/local_collection.js#L640-L642 

2. We should also be sure that if there are callbacks on `CachingChangeObserver`, they don't modify the doc. Therefore we take a deep clone, because in the only case the Meteor code has callbacks here, it modifies the arguments in an unknown way:
https://github.com/meteor/meteor/blob/d072b5b43256005dcb3d96dead5474c1f4893a8d/packages/minimongo/local_collection.js#L1222

Before this, this was always a deep clone, even if there were no callbacks, like here:
https://github.com/meteor/meteor/blob/d072b5b43256005dcb3d96dead5474c1f4893a8d/packages/mongo/observe_multiplex.js#L17-L18

So actually the safety guarantees of CachingChangeObserver have improved , while simultanously reducing the number of deep clones :)